### PR TITLE
Path creator and optimizer for autonomous driving

### PR DESCRIPTION
**Pull request by pauloalbert**
This pull request is part of the work required to make an autonomous robot drive that follows a path. This request is in charge of generating a smooth path, before the robot starts driving. To achieve this,  we included all of the computations in the [Pure pursuit article](https://www.chiefdelphi.com/media/papers/download/5533), each split into methods (details below).

**summed up changelog:**
- Created a class Maintest, that can simulate the calculations on the PC.
- Fixed issues that existed in some of the existing methods.
- Formatted the code.
- Refactored several methods that had inconsistent names, split Vector methods into two types.
- Wrote Javadoc for the methods.

**New methods in the Path class**
- Added `generateFillPoints()`: Takes a Path class of points and fills points at a constant density between each two segments.
- Added `generateSmoothing()`: A smoothing algorithm that takes the new Path and tries to avoid creating sharp turns for the robot.
- Added `generateVelocities()`, `generateCurvature()`, `generateDistance()`: Save in each point useful information.

_note: this code has already been modularly checked (each method was checked independently)._

### CODE REVIEW

File: src/main/java/robot/MainTest.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package robot;
```
Comment (DanKatzuv) [inline src/main/java/robot/MainTest.java]: Why isn't this file inside the `pure_pursuit` package?
```suggestion
package robot.subsystems.drivetrain.pure_pursuit;
```

File: src/main/java/robot/subsystems/drivetrain/pure_pursuit/Path.java [hunk 2]
```diff
@@ -8,7 +8,7 @@
  * @author Lior Barkai
  * This class is the instance of the path, holding the points.
  */
-public class Path {
+public class    Path {
```
Comment (DanKatzuv) [inline src/main/java/robot/subsystems/drivetrain/pure_pursuit/Path.java]: Whitespace is important, but you do not need to have too much of it...
```suggestion
public class  Path {
```

File: src/main/java/robot/subsystems/drivetrain/pure_pursuit/Path.java [hunk 3]
```diff
@@ -154,4 +172,140 @@ public Path copy() {
...
+        this.recursiveDistance(this.length() - 1);
+    }
+
+    /**
+     * Returns the size of the largest length in the list.
+     *
+     * @param i index of current point
+     * @return returns sum of all distances before this point.
+     */
+    private double recursiveDistance(int i) {
```
Comment (DanKatzuv) [inline src/main/java/robot/subsystems/drivetrain/pure_pursuit/Path.java]: Why make this method recursive? [This](https://github.com/DanKatzuv/pure-pursuit/blob/master/path.py#L61-L66) is a simpler solution.

File: src/main/java/robot/subsystems/drivetrain/pure_pursuit/Path.java [hunk 4]
```diff
@@ -154,4 +172,140 @@ public Path copy() {
...
+        if (i == 0)
+            return 0;
+        double d = recursiveDistance(i - 1) + Point.distance(path.get(i), path.get(i - 1));
+        Waypoint p = path.get(i);
+        p.setDistance(d);
+        path.set(i, p);
```
Comment (pauloalbert) [inline src/main/java/robot/subsystems/drivetrain/pure_pursuit/Path.java]: this line is redundant. path.get(i) returns the same pointer, there is no need to save the changes to the waypoint back onto the path.
```suggestion
```

File: src/main/java/robot/subsystems/drivetrain/pure_pursuit/Waypoint.java [hunk 5]
```diff
@@ -47,12 +47,21 @@ public void setSpeed(double speed) {
...
 
-    public double getCurvature() {
-        return curvature;
+    public double getCurvature() { return curvature;}
+
+    public void setCurvature(double curvature) { this.curvature = curvature;}
+
+    @Override
+    public String toString() {
+        return "" + "distance=" + distance + ", \t speed=" + speed + ", \t curvature=" + curvature + ",\t x=" + x + ",\t  y=" + y + "\n";
```
Comment (DanKatzuv) [inline src/main/java/robot/subsystems/drivetrain/pure_pursuit/Waypoint.java]: Change this to a formatted string.

File: src/main/java/robot/subsystems/drivetrain/pure_pursuit/Path.java [hunk 6]
```diff
@@ -154,4 +172,140 @@ public Path copy() {
...
+
+        for (int i = this.length() - 2; i > 0; i--) {
+            this.getWaypoint(i).setSpeed(Math.min(this.getWaypoint(i).getSpeed(), Math.sqrt(Math.pow(this.getWaypoint(i + 1).getSpeed(), 2) + 2 * const_acceleration * Point.distance(this.getWaypoint(i), this.getWaypoint(i-1)))));
+        }
+
+    }
+
+    @Override
+    public String toString() {
+        return "Path{" + "path=" + path + '}';
```
Comment (DanKatzuv) [inline src/main/java/robot/subsystems/drivetrain/pure_pursuit/Path.java]: Change this to a formatted string.

**Pull request merged**